pgpainless-core/src/main/java/org/pgpainless/encryption_signing/EncryptionBuilder.java

```diff
@@ -89,90 +88,54 @@ public ToRecipients onOutputStream(@Nonnull OutputStream outputStream, OpenPgpMe
     class ToRecipientsImpl implements ToRecipients {
 
         @Override
-        public WithAlgorithms toRecipients(@Nonnull PGPPublicKey... keys) {
-            if (keys.length != 0) {
-                List<PGPPublicKey> encryptionKeys = new ArrayList<>();
-                for (PGPPublicKey k : keys) {
+        public WithAlgorithms toRecipients(@Nonnull PGPPublicKeyRing... keys) {
+            if (keys.length == 0) {
+                throw new IllegalArgumentException("No public keys provided.");
+            }
+
+            Map<SubkeyIdentifier, PGPPublicKeyRing> encryptionKeys = new ConcurrentHashMap<>();
```

Comment:
I was reading Robert C. Martins ["Clean Code"](https://enos.itcollege.ee/~jpoial/oop/naited/Clean%20Code.pdf) in which on page 183 he is claiming

> When  Java  was  young,  Doug  Lea  wrote  the  seminal  book Concurrent Programming in Java. Along with the book  he developed several thread-safe collections, which later became part of the JDK in the java.util.concurrent package. The collections in that package are safe for multithreaded situations and they perform well. In fact, the ConcurrentHashMap implementation performs better than HashMap in nearly all situations. It also allows for simultaneous concurrent reads and writes, and it has methods supporting common composite operations that are otherwise not thread safe. If Java 5 is the deployment environment, start with ConcurrentHashMap.

Apparently that is no longer true necessarily, but it stuck with me.
It probably wouldn't hurt reverting to normal HashMaps either.

Reply:
That needs to be checked. But by general means, multithreading support involves locks, and that is unnecessary overhead when you know for sure that the code is not thread-safe. So I'd revert to normal HashMap.